sql-streaming-mqtt/src/main/scala/org/apache/bahir/sql/streaming/mqtt/MQTTStreamSource.scala

```diff
@@ -0,0 +1,191 @@
+      }
+    }
+    client.setCallback(callback)
+    client.connect(mqttConnectOptions)
+    client.subscribe(topic)
+    // It is not possible to initialize offset without `client.connect`
+    offset = fetchLastProcessedOffset()
+    initLock.countDown() // Release.
+  }
+
+  /** Stop this source and free any resources it has allocated. */
+  override def stop(): Unit = {
+    client.disconnect()
+    persistence.close()
+    client.close()
+  }
+
+  /** Returns the maximum available offset for this source. */
+  override def getOffset: Option[Offset] = {
+    if (offset == 0) {
+      None
+    } else {
+      Some(LongOffset(offset))
+    }
+  }
+
+  /**
+   * Returns the data that is between the offsets (`start`, `end`]. When `start` is `None` then
+   * the batch should begin with the first available record. This method must always return the
+   * same data for a particular `start` and `end` pair.
+   */
+  override def getBatch(start: Option[Offset], end: Offset): DataFrame = synchronized {
+    val startIndex = start.getOrElse(LongOffset(0L)).asInstanceOf[LongOffset].offset.toInt
+    val endIndex = end.asInstanceOf[LongOffset].offset.toInt
+    val data: ArrayBuffer[(String, Timestamp)] = ArrayBuffer.empty
+    // Move consumed messages to persistent store.
+    (startIndex + 1 to endIndex).foreach { id =>
+      val element: (String, Timestamp) = messages.getOrElse(id, store.retrieve(id))
+      data += element
+      store.store(id, element)
+      messages.remove(id, element)
+    }
+    log.trace(s"Get Batch invoked, ${data.mkString}")
+    import sqlContext.implicits._
+    data.toDF("value", "timestamp")
+  }
+
+}
+
+class MQTTStreamSourceProvider extends StreamSourceProvider with DataSourceRegister with Logging {
+
+  override def sourceSchema(sqlContext: SQLContext, schema: Option[StructType],
+      providerName: String, parameters: Map[String, String]): (String, StructType) = {
+    ("mqtt", MQTTStreamConstants.SCHEMA_DEFAULT)
+  }
+
+  override def createSource(sqlContext: SQLContext, metadataPath: String,
+      schema: Option[StructType], providerName: String, parameters: Map[String, String]): Source = {
+
+    def e(s: String) = new IllegalArgumentException(s)
+
+    val brokerUrl: String = parameters.getOrElse("brokerUrl", parameters.getOrElse("path",
+      throw e("Please provide a `brokerUrl` by specifying path or .options(\"brokerUrl\",...)")))
+
+
+    val persistence: MqttClientPersistence = parameters.get("persistence") match {
+      case Some("memory") => new MemoryPersistence()
+      case _ => val localStorage: Option[String] = parameters.get("localStorage")
+        localStorage match {
+          case Some(x) => new MqttDefaultFilePersistence(x)
+          case None => new MqttDefaultFilePersistence()
+        }
+    }
+
+    val messageParserWithTimeStamp = (x: Array[Byte]) => (new String(x), Timestamp.valueOf(
+      MQTTStreamConstants.DATE_FORMAT.format(Calendar.getInstance().getTime)))
```

Comment:
An option here would be to use something like Spark DateTimeUtils which handles usage of these classes in ThreadLocal to avoid concurrency issues